File: apps/java/src/test/java/chapter10/FeatureImportanceTest.java
package chapter10;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.entry;
import static org.assertj.core.api.Assertions.within;

import chapter03.DecisionTree;
import chapter03.Leaf;
import java.util.List;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

class FeatureImportanceTest {
  @Nested
  class TreeImportances {
    @Test
    @DisplayName("分割しない木はすべての特徴量の重要度が 0")
    void leaf() {
      var x = Samples.columns("がく片幅", new double[] {0.3, 0.5}, "花弁幅", new double[] {0.1, 0.2});

      assertThat(
              FeatureImportance.treeImportances(new Leaf("setosa"), x, List.of("setosa", "setosa")))
          .containsExactly(entry("がく片幅", 0.0), entry("花弁幅", 0.0));
    }

    @Test
    @DisplayName("1 回だけ分割する木は分割に使った特徴量の重要度が 1")
    void singleSplit() {
      var x =
          Samples.columns(
              "がく片幅", new double[] {0.3, 0.5, 0.4, 0.6}, "花弁幅", new double[] {0.1, 0.2, 0.8, 0.9});
      var t = List.of("setosa", "setosa", "virginica", "virginica");
      var tree = DecisionTree.unlimited().fit(x, t).tree().orElseThrow();

      var importances = FeatureImportance.treeImportances(tree, x, t);

      assertThat(importances).containsKeys("がく片幅", "花弁幅");
      assertThat(importances.get("がく片幅")).isEqualTo(0.0);
      assertThat(importances.get("花弁幅")).isEqualTo(1.0);
    }

    @Test
    @DisplayName("分割で減った不純度を件数で重み付けして割合にする")
    void weighted() {
      var x =
          Samples.columns(
              "花弁長さ",
              new double[] {0.1, 0.2, 0.3, 0.8, 0.7, 0.9},
              "花弁幅",
              new double[] {0.1, 0.1, 0.1, 0.2, 0.9, 0.9});
      var t = List.of("setosa", "setosa", "setosa", "versicolor", "virginica", "virginica");
      var tree = DecisionTree.unlimited().fit(x, t).tree().orElseThrow();

      var importances = FeatureImportance.treeImportances(tree, x, t);

      assertThat(importances.get("花弁長さ")).isCloseTo(7.0 / 11, within(1e-12));
      assertThat(importances.get("花弁幅")).isCloseTo(4.0 / 11, within(1e-12));
    }
  }

  @Nested
  class ForestImportances {
    @Test
    @DisplayName("木が 1 本なら学習に使った行でのその木の重要度と一致する")
    void singleTree() {
      var x =
          Samples.columns(
              "がく片幅",
              new double[] {0.5, 0.3, 0.6, 0.4, 0.5, 0.3, 0.6, 0.4},
              "花弁幅",
              new double[] {0.1, 0.12, 0.14, 0.16, 0.8, 0.82, 0.84, 0.86});
      var t =
          List.of(
              "setosa",
              "setosa",
              "setosa",
              "setosa",
              "virginica",
              "virginica",
              "virginica",
              "virginica");
      var forest = RandomForest.of(1, 2, 0).fit(x, t);
      FittedTree fitted = forest.trees().getFirst();

      var expected =
          FeatureImportance.treeImportances(
              fitted.model().tree().orElseThrow(),
              fitted.rows().stream().map(x::get).toList(),
              fitted.rows().stream().map(t::get).toList());

      assertThat(FeatureImportance.forestImportances(forest, x, t)).isEqualTo(expected);
    }
  }
}

File: apps/dotnet/tests/MachineLearning.Tests/Chapter10/FeatureImportanceTest.fs
module MachineLearning.Tests.Chapter10.FeatureImportanceTest

open Xunit
open MachineLearning.Chapter03
open MachineLearning.Chapter10
open MachineLearning.Chapter10.FeatureImportance

/// 2 つの特徴量の値のリストから、行のリストを作る
let rowsOf (first: string, firstValues: float list) (second: string, secondValues: float list) =
    List.map2 (fun a b -> Map.ofList [ first, a; second, b ]) firstValues secondValues

[<Fact>]
let ``分割しない木はすべての特徴量の重要度が 0`` () =
    let x = rowsOf ("がく片幅", [ 0.3; 0.5 ]) ("花弁幅", [ 0.1; 0.2 ])

    Assert.Equal<Map<string, float>>(
        Map.ofList [ "がく片幅", 0.0; "花弁幅", 0.0 ],
        treeImportances (DecisionTree.Leaf "setosa") x [ "setosa"; "setosa" ]
    )

[<Fact>]
let ``1 回だけ分割する木は分割に使った特徴量の重要度が 1`` () =
    let x = rowsOf ("がく片幅", [ 0.3; 0.5; 0.4; 0.6 ]) ("花弁幅", [ 0.1; 0.2; 0.8; 0.9 ])
    let t = [ "setosa"; "setosa"; "virginica"; "virginica" ]

    let tree = DecisionTree.fit None x t

    Assert.Equal<Map<string, float>>(Map.ofList [ "がく片幅", 0.0; "花弁幅", 1.0 ], treeImportances tree x t)

[<Fact>]
let ``分割で減った不純度を件数で重み付けして割合にする`` () =
    let x =
        rowsOf ("花弁長さ", [ 0.1; 0.2; 0.3; 0.8; 0.7; 0.9 ]) ("花弁幅", [ 0.1; 0.1; 0.2; 0.2; 0.9; 0.9 ])

    let t = [ "setosa"; "setosa"; "setosa"; "versicolor"; "virginica"; "virginica" ]

    let importances = treeImportances (DecisionTree.fit None x t) x t

    Assert.Equal(7.0 / 11.0, importances["花弁長さ"], 12)
    Assert.Equal(4.0 / 11.0, importances["花弁幅"], 12)

[<Fact>]
let ``木が 1 本なら学習に使った行でのその木の重要度と一致する`` () =
    let x =
        rowsOf
            ("がく片幅", [ 0.5; 0.3; 0.6; 0.4; 0.5; 0.3; 0.6; 0.4 ])
            ("花弁幅", [ 0.1; 0.12; 0.14; 0.16; 0.8; 0.82; 0.84; 0.86 ])

    let t = List.replicate 4 "setosa" @ List.replicate 4 "virginica"

    let forest =
        RandomForest.fit
            { RandomForest.defaults with
                NEstimators = 1
                MaxFeatures = 2
            }
            x
            t

    let fitted = List.exactlyOne forest

    let expected =
        treeImportances
            fitted.Tree
            (fitted.Rows |> List.map (fun row -> x[row]))
            (fitted.Rows |> List.map (fun row -> t[row]))

    Assert.Equal<Map<string, float>>(expected, forestImportances forest x t)
